### PR TITLE
Added NodeStates [ACTIVE, SHUTTING_DOWN, SHUTDOWN]

### PR DESCRIPTION
**Pull request by mdogan**
Before there was only an `active` flag in Node. It was set to `false` after partition information check during shutdown. Problem is, during shutdown, before Node itself becomes inactive, all operations are allowed to execute (including operations mutating the data). This can lead to modification of partition data after partition is synced with its backups.

Now, we have one state to separate begin (`SHUTTING_DOWN`) and end (`SHUTDOWN`) of shutdown process. In `SHUTTING_DOWN` state, only certain operations are allowed to run. These operations are (should be) marked with `AllowedDuringShutdown` interface.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java [hunk 1]
```diff
@@ -234,7 +235,7 @@ public Object toObject(Object object) {
 
     @Override
     public boolean isActive() {
-        return node.isActive();
+        return node.getState() == NodeState.ACTIVE;
```
Comment (metanet) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: How about moving this check into node.isActve() method? Then all other node.isActive() checks remain as they are

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: My intention is to make caller aware of the all node states by removing `node.isActive()`, it's not `active/passive` anymore. 

`NodeEngine.isActive()` is called by over SPI level, mostly at that level only important thing is if instance is `ACTIVE` or not. But `Node.getState()` is used on lower levels where difference between `SHUTTING_DOWN` and `SHUTDOWN` is important.

Comment (metanet) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/NodeEngineImpl.java]: ok, thanks for the clarification!

**Comment by metanet** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/instance/NodeState.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * @see Node#start()
+ * @see Node#shutdown(boolean)
+ * @see com.hazelcast.spi.impl.AllowedDuringShutdown
+ *
+ * @since 3.6
+ */
+public enum NodeState {
+
+    /**
+     * Initial state of the Node. An {@code ACTIVE} node is allowed
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/NodeState.java]: is this another change? as currently the node is not active until it's explicitly activated, is it correct?

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/instance/NodeState.java]: no change here. Currently node activated during start process. It's same in this branch too.

Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/instance/NodeState.java]: Before: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/instance/Node.java#L290

After: https://github.com/mdogan/hazelcast/blob/node-state/hazelcast/src/main/java/com/hazelcast/instance/Node.java#L292

**Comment by jerrinot** [comment]
:+1:

**Pull request merged**